sw/device/lib/dif/dif_alert_handler.c

```diff
@@ -4,9 +4,923 @@
 
 #include "sw/device/lib/dif/dif_alert_handler.h"
 
+#include "sw/device/lib/base/bitfield.h"
+
 #include "alert_handler_regs.h"  // Generated.
 
 _Static_assert(ALERT_HANDLER_PARAM_N_CLASSES == 4,
                "Expected four alert classes!");
+_Static_assert(ALERT_HANDLER_PARAM_N_ESC_SEV == 4,
+               "Expected four escalation signals classes!");
+
+dif_alert_handler_result_t dif_alert_handler_init(
+    dif_alert_handler_params_t params, dif_alert_handler_t *handler) {
+  if (handler == NULL) {
+    return kDifAlertHandlerBadArg;
+  }
+
+  // For now, the hardware is hardwired to four signals.
+  if (params.escalation_signal_count != 4) {
+    return kDifAlertHandlerBadArg;
+  }
+
+  handler->params = params;
+
+  return kDifAlertHandlerOk;
+}
+
+/**
+ * Classifies alerts for a single alert class. Returns `false` if any of the
+ * provided configuration is invalid.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool classify_alerts(const dif_alert_handler_t *handler,
```

Comment:
Is distinguishing those failures useful? Note that returning kBadArg is incorrect, since this function may be called after hardware has been touched.

Reply:
Well, because this function feeds in into the public API, you won't be able to return a correct error code to the external caller. I suppose if we don't care about how precise the error codes returned to the user are, we could have just used `bool` as the return type of all the public API.

I personally think that if it fails in several ways, then it should return different error codes, but lets see what the other reviewers think.

Reply:
I think the important question to ask is "how can the caller respond"? I don't think it's useful to report a complicated error to the caller, because it's still going to be hard to figure out where the bad configuration is just from an enum.

In general, I think we've been running on the assumption that all the DIFs can only be usefully debugged with gdb et al.